File: src/mesh/BinaryGrid.js
/**
 * BinaryGrid - Sparse binary voxel storage for region data
 * 
 * Uses compact 16-bit storage per block:
 * - Bits 0-11: Block type ID (0-4095)
 * - Bits 12-15: Fluid level / metadata (0-15)
 * 
 * Memory is allocated sparsely per 16x16x16 section to avoid
 * allocating memory for empty (air) sections.
 */

// Constants
export const SECTION_SIZE = 16;
export const SECTION_VOLUME = SECTION_SIZE * SECTION_SIZE * SECTION_SIZE; // 4096
export const BYTES_PER_BLOCK = 2; // Uint16
export const SECTION_BYTES = SECTION_VOLUME * BYTES_PER_BLOCK; // 8192 bytes per section

// Bit masks for block data
export const BLOCK_ID_MASK = 0x0FFF;     // Bits 0-11 (4096 unique blocks)
export const LEVEL_MASK = 0xF000;         // Bits 12-15 (16 levels)
export const LEVEL_SHIFT = 12;

// Region bounds (Minecraft region = 32x32 chunks)
export const CHUNKS_PER_REGION = 32;
export const REGION_SIZE_XZ = CHUNKS_PER_REGION * SECTION_SIZE; // 512 blocks

// Y bounds (Minecraft 1.18+: -64 to 320)
export const MIN_Y = -64;
export const MAX_Y = 320;
export const Y_SECTIONS = (MAX_Y - MIN_Y) / SECTION_SIZE; // 24 sections

/**
 * Convert world Y to section index
 */
export function worldYToSection(worldY) {
  return Math.floor((worldY - MIN_Y) / SECTION_SIZE);
}

/**
 * Convert section index to world Y base
 */
export function sectionToWorldY(sectionIndex) {
  return sectionIndex * SECTION_SIZE + MIN_Y;
}

/**
 * Create a section key from chunk and section coordinates
 * Uses string key to support arbitrary chunk coordinates across multiple regions
 */
export function makeSectionKey(chunkX, chunkZ, sectionY) {
  return `${chunkX},${chunkZ},${sectionY}`;
}

/**
 * Parse section key back to coordinates
 */
export function parseSectionKey(key) {
  const parts = key.split(',');
  return {
    chunkX: parseInt(parts[0], 10),
    chunkZ: parseInt(parts[1], 10),
    sectionY: parseInt(parts[2], 10),
  };
}

/**
 * Calculate block index within a section
 */
export function blockIndexInSection(localX, localY, localZ) {
  // YZX order (Minecraft standard)
  return localY * SECTION_SIZE * SECTION_SIZE + localZ * SECTION_SIZE + localX;
}

/**
 * BinaryGrid class - sparse storage for a Minecraft region
 */
export class BinaryGrid {
  constructor() {
    // Map of section key -> Uint16Array(4096)
    this.sections = new Map();
    
    // Track bounds for efficient iteration
    this.minChunkX = Infinity;
    this.maxChunkX = -Infinity;
    this.minChunkZ = Infinity;
    this.maxChunkZ = -Infinity;
    this.minSectionY = Infinity;
    this.maxSectionY = -Infinity;
    
    // Statistics
    this.totalBlocks = 0;
    this.sectionCount = 0;
  }
  
  /**
   * Clear all data
   */
  clear() {
    this.sections.clear();
    this.minChunkX = Infinity;
    this.maxChunkX = -Infinity;
    this.minChunkZ = Infinity;
    this.maxChunkZ = -Infinity;
    this.minSectionY = Infinity;
    this.maxSectionY = -Infinity;
    this.totalBlocks = 0;
    this.sectionCount = 0;
  }
  
  /**
   * Get or create a section
   */
  _getOrCreateSection(chunkX, chunkZ, sectionY) {
    const key = makeSectionKey(chunkX, chunkZ, sectionY);
    let section = this.sections.get(key);
    
    if (!section) {
      section = new Uint16Array(SECTION_VOLUME);
      this.sections.set(key, section);
      this.sectionCount++;
      
      // Update bounds
      if (chunkX < this.minChunkX) this.minChunkX = chunkX;
      if (chunkX > this.maxChunkX) this.maxChunkX = chunkX;
      if (chunkZ < this.minChunkZ) this.minChunkZ = chunkZ;
      if (chunkZ > this.maxChunkZ) this.maxChunkZ = chunkZ;
      if (sectionY < this.minSectionY) this.minSectionY = sectionY;
      if (sectionY > this.maxSectionY) this.maxSectionY = sectionY;
    }
    
    return section;
  }
  
  /**
   * Get a section (returns undefined if not exists)
   */
  getSection(chunkX, chunkZ, sectionY) {
    const key = makeSectionKey(chunkX, chunkZ, sectionY);
    return this.sections.get(key);
  }
  
  /**
   * Set a block at world coordinates
   * @param {number} worldX - World X coordinate
   * @param {number} worldY - World Y coordinate
   * @param {number} worldZ - World Z coordinate
   * @param {number} blockId - Block type ID (0-4095)
   * @param {number} level - Fluid level or metadata (0-15)
   */
  setBlock(worldX, worldY, worldZ, blockId, level = 0) {
    // Convert to chunk/section coordinates
    const chunkX = Math.floor(worldX / SECTION_SIZE);
    const chunkZ = Math.floor(worldZ / SECTION_SIZE);
    const sectionY = worldYToSection(worldY);
    
    // Local coordinates within section
    const localX = ((worldX % SECTION_SIZE) + SECTION_SIZE) % SECTION_SIZE;
    const localY = ((worldY - MIN_Y) % SECTION_SIZE + SECTION_SIZE) % SECTION_SIZE;
    const localZ = ((worldZ % SECTION_SIZE) + SECTION_SIZE) % SECTION_SIZE;
    
    // Get or create section
    const section = this._getOrCreateSection(chunkX, chunkZ, sectionY);
    
    // Calculate index and set value
    const index = blockIndexInSection(localX, localY, localZ);
    const oldValue = section[index];
    const newValue = (blockId & BLOCK_ID_MASK) | ((level & 0xF) << LEVEL_SHIFT);
    
    section[index] = newValue;
    
    // Update block count
    if (oldValue === 0 && newValue !== 0) {
      this.totalBlocks++;
    } else if (oldValue !== 0 && newValue === 0) {
      this.totalBlocks--;
    }
  }
  
  /**
   * Set a block using chunk-relative coordinates
   * @param {number} chunkX - Chunk X coordinate
   * @param {number} chunkZ - Chunk Z coordinate
   * @param {number} localX - Local X within chunk (0-15)
   * @param {number} worldY - World Y coordinate
   * @param {number} localZ - Local Z within chunk (0-15)
   * @param {number} blockId - Block type ID (0-4095)
   * @param {number} level - Fluid level or metadata (0-15)
   */
  setBlockLocal(chunkX, chunkZ, localX, worldY, localZ, blockId, level = 0) {
    const sectionY = worldYToSection(worldY);
    const localY = ((worldY - MIN_Y) % SECTION_SIZE + SECTION_SIZE) % SECTION_SIZE;
    
    const section = this._getOrCreateSection(chunkX, chunkZ, sectionY);
    const index = blockIndexInSection(localX, localY, localZ);
    const oldValue = section[index];
    const newValue = (blockId & BLOCK_ID_MASK) | ((level & 0xF) << LEVEL_SHIFT);
    
    section[index] = newValue;
    
    if (oldValue === 0 && newValue !== 0) {
      this.totalBlocks++;
    } else if (oldValue !== 0 && newValue === 0) {
      this.totalBlocks--;
    }
  }
  
  /**
   * FAST: Set a block with pre-computed section - for batch decoding
   * @param {Uint16Array} section - Pre-fetched section array
   * @param {number} localX - Local X (0-15)
   * @param {number} localY - Local Y (0-15)
   * @param {number} localZ - Local Z (0-15)
   * @param {number} blockId - Block ID (0-4095)
   * @param {number} level - Fluid level (0-15)
   */
  setBlockInSection(section, localX, localY, localZ, blockId, level = 0) {
    const index = localY * SECTION_SIZE * SECTION_SIZE + localZ * SECTION_SIZE + localX;
    const oldValue = section[index];
    const newValue = (blockId & BLOCK_ID_MASK) | ((level & 0xF) << LEVEL_SHIFT);
    section[index] = newValue;
    
    if (oldValue === 0 && newValue !== 0) {
      this.totalBlocks++;
    } else if (oldValue !== 0 && newValue === 0) {
      this.totalBlocks--;
    }
  }
  
  /**
   * Get block data at world coordinates
   * @returns {{blockId: number, level: number} | null}
   */
  getBlock(worldX, worldY, worldZ) {
    const chunkX = Math.floor(worldX / SECTION_SIZE);
    const chunkZ = Math.floor(worldZ / SECTION_SIZE);
    const sectionY = worldYToSection(worldY);
    
    const section = this.getSection(chunkX, chunkZ, sectionY);
    if (!section) return null;
    
    const localX = ((worldX % SECTION_SIZE) + SECTION_SIZE) % SECTION_SIZE;
    const localY = ((worldY - MIN_Y) % SECTION_SIZE + SECTION_SIZE) % SECTION_SIZE;
    const localZ = ((worldZ % SECTION_SIZE) + SECTION_SIZE) % SECTION_SIZE;
    
    const index = blockIndexInSection(localX, localY, localZ);
    const value = section[index];
    
    return {
      blockId: value & BLOCK_ID_MASK,
      level: (value & LEVEL_MASK) >> LEVEL_SHIFT,
    };
  }
  
  /**
   * Get block ID at world coordinates (fast path)
   * @returns {number} Block ID or 0 if air/missing
   */
  getBlockId(worldX, worldY, worldZ) {
    const chunkX = Math.floor(worldX / SECTION_SIZE);
    const chunkZ = Math.floor(worldZ / SECTION_SIZE);
    const sectionY = worldYToSection(worldY);
    
    const section = this.getSection(chunkX, chunkZ, sectionY);
    if (!section) return 0;
    
    const localX = ((worldX % SECTION_SIZE) + SECTION_SIZE) % SECTION_SIZE;
    const localY = ((worldY - MIN_Y) % SECTION_SIZE + SECTION_SIZE) % SECTION_SIZE;
    const localZ = ((worldZ % SECTION_SIZE) + SECTION_SIZE) % SECTION_SIZE;
    
    const index = blockIndexInSection(localX, localY, localZ);
    return section[index] & BLOCK_ID_MASK;
  }
  
  /**
   * Get fluid level at world coordinates
   * @returns {number} Level 0-15 or -1 if not a fluid
   */
  getLevel(worldX, worldY, worldZ) {
    const chunkX = Math.floor(worldX / SECTION_SIZE);
    const chunkZ = Math.floor(worldZ / SECTION_SIZE);
    const sectionY = worldYToSection(worldY);
    
    const section = this.getSection(chunkX, chunkZ, sectionY);
    if (!section) return -1;
    
    const localX = ((worldX % SECTION_SIZE) + SECTION_SIZE) % SECTION_SIZE;
    const localY = ((worldY - MIN_Y) % SECTION_SIZE + SECTION_SIZE) % SECTION_SIZE;
    const localZ = ((worldZ % SECTION_SIZE) + SECTION_SIZE) % SECTION_SIZE;
    
    const index = blockIndexInSection(localX, localY, localZ);
    const value = section[index];
    
    if (value === 0) return -1;
    return (value & LEVEL_MASK) >> LEVEL_SHIFT;
  }
  
  /**
   * Check if a position has a non-air block
   */
  hasBlock(worldX, worldY, worldZ) {
    return this.getBlockId(worldX, worldY, worldZ) !== 0;
  }
  
  /**
   * Get all section keys
   * @returns {number[]} Array of section keys
   */
  getSectionKeys() {
    return Array.from(this.sections.keys());
  }
  
  /**
   * Iterate over all non-empty sections
   * @param {Function} callback - (section: Uint16Array, chunkX, chunkZ, sectionY) => void
   */
  forEachSection(callback) {
    for (const [key, section] of this.sections) {
      const { chunkX, chunkZ, sectionY } = parseSectionKey(key);
      callback(section, chunkX, chunkZ, sectionY);
    }
  }
  
  /**
   * Iterate over all non-air blocks in a section
   * @param {Function} callback - (blockId, level, localX, localY, localZ) => void
   */
  forEachBlockInSection(section, callback) {
    for (let i = 0; i < SECTION_VOLUME; i++) {
      const value = section[i];
      if (value === 0) continue;
      
      const localX = i % SECTION_SIZE;
      const localZ = Math.floor(i / SECTION_SIZE) % SECTION_SIZE;
      const localY = Math.floor(i / (SECTION_SIZE * SECTION_SIZE));
      
      const blockId = value & BLOCK_ID_MASK;
      const level = (value & LEVEL_MASK) >> LEVEL_SHIFT;
      
      callback(blockId, level, localX, localY, localZ);
    }
  }
  
  /**
   * Get raw section data for a chunk column
   * Returns all sections for a single chunk (16x384x16)
   * @returns {Map<number, Uint16Array>} sectionY -> section data
   */
  getChunkColumn(chunkX, chunkZ) {
    const result = new Map();
    
    for (let sectionY = 0; sectionY < Y_SECTIONS; sectionY++) {
      const section = this.getSection(chunkX, chunkZ, sectionY);
      if (section) {
        result.set(sectionY, section);
      }
    }
    
    return result;
  }
  
  /**
   * Get memory usage estimate
   */
  getMemoryUsage() {
    return {
      sectionCount: this.sectionCount,
      bytesUsed: this.sectionCount * SECTION_BYTES,
      totalBlocks: this.totalBlocks,
    };
  }
  
  /**
   * Get grid bounds in world coordinates
   */
  getBounds() {
    if (this.sectionCount === 0) {
      return null;
    }
    
    return {
      minX: this.minChunkX * SECTION_SIZE,
      maxX: (this.maxChunkX + 1) * SECTION_SIZE - 1,
      minY: sectionToWorldY(this.minSectionY),
      maxY: sectionToWorldY(this.maxSectionY + 1) - 1,
      minZ: this.minChunkZ * SECTION_SIZE,
      maxZ: (this.maxChunkZ + 1) * SECTION_SIZE - 1,
    };
  }
  
  /**
   * Export grid data for worker transfer
   * Returns data that can be sent to workers via postMessage
   */
  export() {
    const sectionsData = [];
    
    for (const [key, section] of this.sections) {
      sectionsData.push({
        key,
        data: section.buffer.slice(0), // Copy buffer
      });
    }
    
    return {
      sections: sectionsData,
      bounds: {
        minChunkX: this.minChunkX,
        maxChunkX: this.maxChunkX,
        minChunkZ: this.minChunkZ,
        maxChunkZ: this.maxChunkZ,
        minSectionY: this.minSectionY,
        maxSectionY: this.maxSectionY,
      },
      totalBlocks: this.totalBlocks,
    };
  }
  
  /**
   * Import grid data from export
   */
  static import(data) {
    const grid = new BinaryGrid();
    
    for (const { key, data: buffer } of data.sections) {
      const section = new Uint16Array(buffer);
      grid.sections.set(key, section);
    }
    
    grid.minChunkX = data.bounds.minChunkX;
    grid.maxChunkX = data.bounds.maxChunkX;
    grid.minChunkZ = data.bounds.minChunkZ;
    grid.maxChunkZ = data.bounds.maxChunkZ;
    grid.minSectionY = data.bounds.minSectionY;
    grid.maxSectionY = data.bounds.maxSectionY;
    grid.totalBlocks = data.totalBlocks;
    grid.sectionCount = data.sections.length;
    
    return grid;
  }
}

export default BinaryGrid;

File: src/wasm-mesher/src/grid/binary_grid.rs
//! BinaryGrid - Sparse voxel storage for block data
//!
//! Uses 16-bit values per block:
//! - Bits 0-11: Block ID (4096 unique blocks)
//! - Bits 12-15: Fluid level / metadata

use std::collections::HashMap;
use crate::types::{
    SectionKey, SECTION_SIZE, SECTION_VOLUME, BLOCK_ID_MASK, LEVEL_MASK, LEVEL_SHIFT,
    SLAB_MASK, SLAB_SHIFT, block_index_in_section, world_y_to_section,
};

/// Sparse binary grid for block storage
pub struct BinaryGrid {
    /// Sections stored by packed key
    sections: HashMap<u64, Box<[u16; SECTION_VOLUME]>>,
}

impl BinaryGrid {
    pub fn new() -> Self {
        Self {
            sections: HashMap::new(),
        }
    }

    /// Import from serialized byte data
    /// Format: [num_sections: u32][section_key: u64, data: [u16; 4096]]...
    pub fn from_bytes(data: &[u8]) -> Self {
        let mut grid = Self::new();
        
        if data.len() < 4 {
            return grid;
        }

        let num_sections = u32::from_le_bytes([data[0], data[1], data[2], data[3]]) as usize;
        let mut offset = 4;

        for _ in 0..num_sections {
            if offset + 8 + SECTION_VOLUME * 2 > data.len() {
                break;
            }

            // Read section key (packed u64)
            let key = u64::from_le_bytes([
                data[offset], data[offset + 1], data[offset + 2], data[offset + 3],
                data[offset + 4], data[offset + 5], data[offset + 6], data[offset + 7],
            ]);
            offset += 8;

            // Read section data
            let mut section = Box::new([0u16; SECTION_VOLUME]);
            for i in 0..SECTION_VOLUME {
                section[i] = u16::from_le_bytes([
                    data[offset + i * 2],
                    data[offset + i * 2 + 1],
                ]);
            }
            offset += SECTION_VOLUME * 2;

            grid.sections.insert(key, section);
        }

        grid
    }

    /// Get a section by key
    #[inline]
    pub fn get_section(&self, key: &SectionKey) -> Option<&[u16; SECTION_VOLUME]> {
        self.sections.get(&key.to_packed()).map(|s| s.as_ref())
    }

    /// Get a section by packed key
    #[inline]
    pub fn get_section_by_packed(&self, packed: u64) -> Option<&[u16; SECTION_VOLUME]> {
        self.sections.get(&packed).map(|s| s.as_ref())
    }

    /// Get block value at world coordinates
    #[inline]
    pub fn get_block(&self, world_x: i32, world_y: i32, world_z: i32) -> u16 {
        let chunk_x = world_x.div_euclid(SECTION_SIZE as i32);
        let chunk_z = world_z.div_euclid(SECTION_SIZE as i32);
        let section_y = world_y_to_section(world_y);

        let local_x = world_x.rem_euclid(SECTION_SIZE as i32) as usize;
        let local_y = world_y.rem_euclid(SECTION_SIZE as i32) as usize;
        let local_z = world_z.rem_euclid(SECTION_SIZE as i32) as usize;

        let key = SectionKey::new(chunk_x, chunk_z, section_y);
        if let Some(section) = self.get_section(&key) {
            let idx = block_index_in_section(local_x, local_y, local_z);
            section[idx]
        } else {
            0 // Air
        }
    }

    /// Get block ID at world coordinates
    #[inline]
    pub fn get_block_id(&self, world_x: i32, world_y: i32, world_z: i32) -> u16 {
        self.get_block(world_x, world_y, world_z) & BLOCK_ID_MASK
    }

    /// Get fluid level at world coordinates
    #[inline]
    pub fn get_level(&self, world_x: i32, world_y: i32, world_z: i32) -> u8 {
        ((self.get_block(world_x, world_y, world_z) & LEVEL_MASK) >> LEVEL_SHIFT) as u8
    }

    /// Get slab type at world coordinates
    #[inline]
    pub fn get_slab_type(&self, world_x: i32, world_y: i32, world_z: i32) -> u8 {
        ((self.get_block(world_x, world_y, world_z) & SLAB_MASK) >> SLAB_SHIFT) as u8
    }

    /// Iterate over all sections
    pub fn iter_sections(&self) -> impl Iterator<Item = (SectionKey, &[u16; SECTION_VOLUME])> {
        self.sections.iter().map(|(packed, section)| {
            (SectionKey::from_packed(*packed), section.as_ref())
        })
    }

    /// Get or create a section for writing
    /// Returns a mutable reference to the section data
    pub fn get_or_create_section(&mut self, chunk_x: i32, chunk_z: i32, section_y: i32) -> &mut [u16; SECTION_VOLUME] {
        let key = SectionKey::new(chunk_x, chunk_z, section_y);
        let packed = key.to_packed();
        
        self.sections.entry(packed)
            .or_insert_with(|| Box::new([0u16; SECTION_VOLUME]))
            .as_mut()
    }

    /// Get a mutable section by key (for internal use)
    #[inline]
    pub fn get_section_mut(&mut self, key: &SectionKey) -> Option<&mut [u16; SECTION_VOLUME]> {
        self.sections.get_mut(&key.to_packed()).map(|s| s.as_mut())
    }

    /// Get neighbor section for a given section key and direction
    #[inline]
    pub fn get_neighbor_section(
        &self,
        key: &SectionKey,
        dx: i32,
        dy: i32,
        dz: i32,
    ) -> Option<&[u16; SECTION_VOLUME]> {
        let neighbor_key = SectionKey::new(
            key.chunk_x + dx,
            key.chunk_z + dz,
            key.section_y + dy,
        );
        self.get_section(&neighbor_key)
    }

    /// Merge another grid into this one (copies all sections)
    pub fn merge_from(&mut self, other: &BinaryGrid) {
        for (packed, section) in &other.sections {
            self.sections.insert(*packed, section.clone());
        }
    }
}

impl Default for BinaryGrid {
    fn default() -> Self {
        Self::new()
    }
}

